File: auth-lib/src/main/java/com/spotify/sdk/android/auth/AuthorizationRequest.java
/*
 * Copyright (c) 2015-2016 Spotify AB
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.sdk.android.auth;

import static com.spotify.sdk.android.auth.AccountsQueryParameters.ASSOCIATED_CONTENT;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import org.json.JSONObject;

import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Map;

/**
 * An object that helps construct the request that is sent to Spotify authorization service.
 * To create one use {@link AuthorizationRequest.Builder}
 *
 * @see <a href="https://developer.spotify.com/web-api/authorization-guide">Web API Authorization guide</a>
 */
public class AuthorizationRequest implements Parcelable {

    static final String ACCOUNTS_SCHEME = "https";
    static final String ACCOUNTS_AUTHORITY = "accounts.spotify.com";
    static final String ACCOUNTS_PATH = "authorize";
    static final String SCOPES_SEPARATOR = " ";
    @VisibleForTesting
    public static final String SPOTIFY_SDK = "spotify-sdk";
    @VisibleForTesting
    public static final String ANDROID_SDK = "android-sdk";
    private static final String KEY_URI = "uri";
    private static final String KEY_URL = "url";

    private final String mClientId;
    private final String mResponseType;
    private final String mRedirectUri;
    private final String mState;
    private final String[] mScopes;
    private final boolean mShowDialog;
    private final Map<String, String> mCustomParams;
    private final String mCampaign;
    private final String mContentUri;
    private final String mContentUrl;

    /**
     * Use this builder to create an {@link AuthorizationRequest}
     *
     * @see AuthorizationRequest
     */
    public static class Builder {

        private final String mClientId;
        private final AuthorizationResponse.Type mResponseType;
        private final String mRedirectUri;

        private String mState;
        private String[] mScopes;
        private boolean mShowDialog;
        private String mCampaign;
        private final Map<String, String> mCustomParams = new HashMap<>();
        private String mContentUri;
        private String mContentUrl;

        public Builder(String clientId, AuthorizationResponse.Type responseType, String redirectUri) {
            if (clientId == null) {
                throw new IllegalArgumentException("Client ID can't be null");
            }
            if (responseType == null) {
                throw new IllegalArgumentException("Response type can't be null");
            }
            if (redirectUri == null || redirectUri.length() == 0) {
                throw new IllegalArgumentException("Redirect URI can't be null or empty");
            }

            mClientId = clientId;
            mResponseType = responseType;
            mRedirectUri = redirectUri;
        }

        public Builder setState(String state) {
            mState = state;
            return this;
        }

        public Builder setScopes(String[] scopes) {
            mScopes = scopes;
            return this;
        }

        public Builder setShowDialog(boolean showDialog) {
            mShowDialog = showDialog;
            return this;
        }

        public Builder setCustomParam(String key, String value) {
            if (key == null || key.isEmpty()) {
                throw new IllegalArgumentException("Custom parameter key can't be null or empty");
            }
            if (value == null || value.isEmpty()) {
                throw new IllegalArgumentException("Custom parameter value can't be null or empty");
            }
            mCustomParams.put(key, value);
            return this;
        }

        public Builder setCampaign(String campaign) {
            mCampaign = campaign;
            return this;
        }

        public Builder setContentUri(String contentUri) {
            mContentUri = contentUri;
            return this;
        }

        public Builder setContentUrl(String contentUrl) {
            mContentUrl = contentUrl;
            return this;
        }

        public AuthorizationRequest build() {
            return new AuthorizationRequest(mClientId, mResponseType, mRedirectUri,
                    mState, mScopes, mShowDialog, mCustomParams, mCampaign, mContentUri, mContentUrl);
        }
    }

    public AuthorizationRequest(Parcel source) {
        mClientId = source.readString();
        mResponseType = source.readString();
        mRedirectUri = source.readString();
        mState = source.readString();
        mScopes = source.createStringArray();
        mShowDialog = source.readByte() == 1;
        mCustomParams = new HashMap<>();
        mCampaign = source.readString();
        mContentUri = source.readString();
        mContentUrl = source.readString();
        Bundle bundle = source.readBundle(getClass().getClassLoader());
        for (String key : bundle.keySet()) {
            mCustomParams.put(key, bundle.getString(key));
        }
    }

    public String getClientId() {
        return mClientId;
    }

    public String getResponseType() {
        return mResponseType;
    }

    public String getRedirectUri() {
        return mRedirectUri;
    }

    public String getState() {
        return mState;
    }

    public String[] getScopes() {
        return mScopes;
    }

    public String getCustomParam(String key) {
        return mCustomParams.get(key);
    }

    public String getEncodedContent() {
        JSONObject contentJson = getContentsJson();

        if(contentJson.length() == 0) {
            return null; // No content to encode
        }

        return Base64.encodeToString(contentJson.toString().getBytes(Charset.forName("UTF-8")), Base64.URL_SAFE | Base64.NO_WRAP);
    }

    @NonNull
    private JSONObject getContentsJson() {
        JSONObject contentJson = new JSONObject();

        if(mContentUri != null && !mContentUri.isEmpty()) {
            try {
                contentJson.put(KEY_URI, mContentUri);
            } catch (Exception e) {
                throw new IllegalArgumentException("Error creating JSON for content URI: " + e.getMessage());
            }
        }

        if(mContentUrl != null && !mContentUrl.isEmpty()) {
            try {
                contentJson.put(KEY_URL, mContentUrl);
            } catch (Exception e) {
                throw new IllegalArgumentException("Error creating JSON for content URL: " + e.getMessage());
            }
        }
        return contentJson;
    }

    @NonNull
    public String getCampaign() { return TextUtils.isEmpty(mCampaign) ? ANDROID_SDK : mCampaign; }

    @NonNull
    public String getSource() { return SPOTIFY_SDK; }

    @NonNull
    public String getMedium() { return ANDROID_SDK; }

    private AuthorizationRequest(String clientId,
                                 AuthorizationResponse.Type responseType,
                                 String redirectUri,
                                 String state,
                                 String[] scopes,
                                 boolean showDialog,
                                 Map<String, String> customParams,
                                 String campaign,
                                 String contentUri,
                                 String contentUrl
                                 ) {

        mClientId = clientId;
        mResponseType = responseType.toString();
        mRedirectUri = redirectUri;
        mState = state;
        mScopes = scopes;
        mShowDialog = showDialog;
        mCustomParams = customParams;
        mCampaign = campaign;
        mContentUri = contentUri;
        mContentUrl = contentUrl;
    }

    public Uri toUri() {
        Uri.Builder uriBuilder = new Uri.Builder();
        uriBuilder.scheme(ACCOUNTS_SCHEME)
                .authority(ACCOUNTS_AUTHORITY)
                .appendPath(ACCOUNTS_PATH)
                .appendQueryParameter(AccountsQueryParameters.CLIENT_ID, mClientId)
                .appendQueryParameter(AccountsQueryParameters.RESPONSE_TYPE, mResponseType)
                .appendQueryParameter(AccountsQueryParameters.REDIRECT_URI, mRedirectUri)
                .appendQueryParameter(AccountsQueryParameters.SHOW_DIALOG, String.valueOf(mShowDialog))
                .appendQueryParameter(AccountsQueryParameters.UTM_SOURCE, getSource())
                .appendQueryParameter(AccountsQueryParameters.UTM_MEDIUM, getMedium())
                .appendQueryParameter(AccountsQueryParameters.UTM_CAMPAIGN, getCampaign());

        if (mScopes != null && mScopes.length > 0) {
            uriBuilder.appendQueryParameter(AccountsQueryParameters.SCOPE, scopesToString());
        }

        if (mState != null) {
            uriBuilder.appendQueryParameter(AccountsQueryParameters.STATE, mState);
        }

        if (mCustomParams.size() > 0) {
            for (Map.Entry<String, String> entry : mCustomParams.entrySet()) {
                uriBuilder.appendQueryParameter(entry.getKey(), entry.getValue());
            }
        }

        String associatedContent = getEncodedContent();

        if(associatedContent != null) {
            uriBuilder.appendQueryParameter(ASSOCIATED_CONTENT, associatedContent);
        }

        return uriBuilder.build();
    }

    private String scopesToString() {
        StringBuilder concatScopes = new StringBuilder();
        for (String scope : mScopes) {
            concatScopes.append(scope);
            concatScopes.append(SCOPES_SEPARATOR);
        }
        return concatScopes.toString().trim();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mClientId);
        dest.writeString(mResponseType);
        dest.writeString(mRedirectUri);
        dest.writeString(mState);
        dest.writeStringArray(mScopes);
        dest.writeByte((byte) (mShowDialog ? 1 : 0));
        dest.writeString(mCampaign);
        dest.writeString(mContentUri);
        dest.writeString(mContentUrl);

        Bundle bundle = new Bundle();
        for (Map.Entry<String, String> entry : mCustomParams.entrySet()) {
            bundle.putString(entry.getKey(), entry.getValue());
        }
        dest.writeBundle(bundle);
    }

    public static final Parcelable.Creator<AuthorizationRequest> CREATOR = new Parcelable.Creator<AuthorizationRequest>() {

        @Override
        public AuthorizationRequest createFromParcel(Parcel source) {
            return new AuthorizationRequest(source);
        }

        @Override
        public AuthorizationRequest[] newArray(int size) {
            return new AuthorizationRequest[size];
        }
    };
}

File: auth-lib/src/main/kotlin/com/spotify/sdk/android/auth/AuthorizationRequest.kt
/*
 * Copyright (c) 2015-2016 Spotify AB
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.sdk.android.auth

import android.net.Uri
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import android.text.TextUtils
import androidx.annotation.VisibleForTesting

/**
 * An object that helps construct the request that is sent to Spotify authorization service.
 * To create one use [AuthorizationRequest.Builder]
 *
 * @see <a href="https://developer.spotify.com/web-api/authorization-guide">Web API Authorization guide</a>
 */
data class AuthorizationRequest internal constructor(
    val clientId: String,
    val responseType: String,
    val redirectUri: String,
    val state: String?,
    val scopes: Array<String>?,
    val showDialog: Boolean,
    val customParams: Map<String, String>,
    private val campaign: String?,
    val pkceInformation: PKCEInformation?
) : Parcelable {

    @Suppress("DEPRECATION")
    constructor(source: Parcel) : this(
        clientId = source.readString() ?: throw IllegalStateException("clientId cannot be null in parcel"),
        responseType = source.readString() ?: throw IllegalStateException("responseType cannot be null in parcel"),
        redirectUri = source.readString() ?: throw IllegalStateException("redirectUri cannot be null in parcel"),
        state = source.readString(),
        scopes = source.createStringArray(),
        showDialog = source.readByte() == 1.toByte(),
        customParams = HashMap(),
        campaign = source.readString(),
        pkceInformation = source.readParcelable(PKCEInformation::class.java.classLoader)
    ) {
        val bundle = source.readBundle(javaClass.classLoader) ?: Bundle()
        for (key in bundle.keySet()) {
            bundle.getString(key)?.let { value ->
                (customParams as HashMap)[key] = value
            }
        }
    }

    fun getCustomParam(key: String): String? {
        return customParams[key]
    }

    fun getCampaign(): String {
        return campaign?.takeUnless { TextUtils.isEmpty(it) } ?: ANDROID_SDK
    }

    fun getSource(): String = SPOTIFY_SDK

    fun getMedium(): String = ANDROID_SDK

    fun toUri(): Uri {
        val uriBuilder = Uri.Builder()
        uriBuilder.scheme(ACCOUNTS_SCHEME)
            .authority(ACCOUNTS_AUTHORITY)
            .appendPath(ACCOUNTS_PATH)
            .appendQueryParameter(AccountsQueryParameters.CLIENT_ID, clientId)
            .appendQueryParameter(AccountsQueryParameters.RESPONSE_TYPE, responseType)
            .appendQueryParameter(AccountsQueryParameters.REDIRECT_URI, redirectUri)
            .appendQueryParameter(AccountsQueryParameters.SHOW_DIALOG, showDialog.toString())
            .appendQueryParameter(AccountsQueryParameters.UTM_SOURCE, getSource())
            .appendQueryParameter(AccountsQueryParameters.UTM_MEDIUM, getMedium())
            .appendQueryParameter(AccountsQueryParameters.UTM_CAMPAIGN, getCampaign())

        if (scopes != null && scopes.isNotEmpty()) {
            uriBuilder.appendQueryParameter(AccountsQueryParameters.SCOPE, scopesToString())
        }

        if (state != null) {
            uriBuilder.appendQueryParameter(AccountsQueryParameters.STATE, state)
        }

        if (customParams.isNotEmpty()) {
            for ((key, value) in customParams) {
                uriBuilder.appendQueryParameter(key, value)
            }
        }

        if (pkceInformation != null) {
            uriBuilder.appendQueryParameter(AccountsQueryParameters.CODE_CHALLENGE, pkceInformation.challenge)
            uriBuilder.appendQueryParameter(AccountsQueryParameters.CODE_CHALLENGE_METHOD, pkceInformation.codeChallengeMethod)
        }

        return uriBuilder.build()
    }

    private fun scopesToString(): String {
        if (scopes == null) return ""
        val concatScopes = StringBuilder()
        for (scope in scopes) {
            concatScopes.append(scope)
            concatScopes.append(SCOPES_SEPARATOR)
        }
        return concatScopes.toString().trim()
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(clientId)
        dest.writeString(responseType)
        dest.writeString(redirectUri)
        dest.writeString(state)
        dest.writeStringArray(scopes)
        dest.writeByte(if (showDialog) 1 else 0)
        dest.writeString(campaign)
        dest.writeParcelable(pkceInformation, flags)

        val bundle = Bundle()
        for ((key, value) in customParams) {
            bundle.putString(key, value)
        }
        dest.writeBundle(bundle)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as AuthorizationRequest

        if (clientId != other.clientId) return false
        if (responseType != other.responseType) return false
        if (redirectUri != other.redirectUri) return false
        if (state != other.state) return false
        if (scopes != null) {
            if (other.scopes == null) return false
            if (!scopes.contentEquals(other.scopes)) return false
        } else if (other.scopes != null) return false
        if (showDialog != other.showDialog) return false
        if (customParams != other.customParams) return false
        if (campaign != other.campaign) return false
        if (pkceInformation != other.pkceInformation) return false

        return true
    }

    override fun hashCode(): Int {
        var result = clientId.hashCode()
        result = 31 * result + responseType.hashCode()
        result = 31 * result + redirectUri.hashCode()
        result = 31 * result + (state?.hashCode() ?: 0)
        result = 31 * result + (scopes?.contentHashCode() ?: 0)
        result = 31 * result + showDialog.hashCode()
        result = 31 * result + customParams.hashCode()
        result = 31 * result + (campaign?.hashCode() ?: 0)
        result = 31 * result + (pkceInformation?.hashCode() ?: 0)
        return result
    }

    /**
     * Use this builder to create an [AuthorizationRequest]
     *
     * @see AuthorizationRequest
     */
    class Builder(
        private val clientId: String,
        private val responseType: AuthorizationResponse.Type,
        private val redirectUri: String
    ) {
        private var state: String? = null
        private var scopes: Array<String>? = null
        private var showDialog: Boolean = false
        private var campaign: String? = null
        private var pkceInformation: PKCEInformation? = null
        private val customParams = HashMap<String, String>()

        init {
            require(redirectUri.isNotEmpty()) { "Redirect URI cannot be empty" }
        }

        fun setState(state: String?) = apply {
            this.state = state
        }

        fun setScopes(scopes: Array<String>?) = apply {
            this.scopes = scopes
        }

        fun setShowDialog(showDialog: Boolean) = apply {
            this.showDialog = showDialog
        }

        fun setCustomParam(key: String, value: String) = apply {
            require(key.isNotEmpty()) { "Custom parameter key cannot be empty" }
            require(value.isNotEmpty()) { "Custom parameter value cannot be empty" }
            customParams[key] = value
        }

        fun setCampaign(campaign: String?) = apply {
            this.campaign = campaign
        }

        fun setPkceInformation(pkceInformation: PKCEInformation?) = apply {
            this.pkceInformation = pkceInformation
        }

        fun build(): AuthorizationRequest {
            return AuthorizationRequest(
                clientId,
                responseType.toString(),
                redirectUri,
                state,
                scopes,
                showDialog,
                customParams,
                campaign,
                pkceInformation
            )
        }
    }

    companion object {
        const val ACCOUNTS_SCHEME = "https"
        const val ACCOUNTS_AUTHORITY = "accounts.spotify.com"
        const val ACCOUNTS_PATH = "authorize"
        const val SCOPES_SEPARATOR = " "
        @VisibleForTesting
        const val SPOTIFY_SDK = "spotify-sdk"
        @VisibleForTesting
        const val ANDROID_SDK = "android-sdk"

        @JvmField
        val CREATOR = object : Parcelable.Creator<AuthorizationRequest> {
            override fun createFromParcel(source: Parcel): AuthorizationRequest {
                return AuthorizationRequest(source)
            }

            override fun newArray(size: Int): Array<AuthorizationRequest?> {
                return arrayOfNulls(size)
            }
        }
    }
}
